### PR TITLE
Embedding developments: New feature electron embedding and bugfixes

### PR DESCRIPTION
**Pull request by perahrens**
Presented in RECO AT 14.07.2017.

- New feature to enable electron embedding:
Boolean parameter switchToMuonEmbedding replaced by integer parameter particleToEmbed representing the PDGID of the particle to be embedded.

- Workaround to keep prunedGenParticles:
New customizer function to keep genParticle collections that are removed in the PAT step on data.

- Bugfix in EmbeddingLHEProducer:
Only tau-leptons were added to the LHE product. Now crosschecking with list of allowed particles.

- Bugfix in Merge Collections (found after presentation):
First merge the two collections, then save the new collection.

**Testcases:**
Clone the following repository into your CMSSW_9_2_3patch2 src folder:
`git clone https://github.com/perahrens/gc_configs_for_embedding.git`

It contains three test configuration folders:
1. ElEl_data_2017_CMSSW923p2_ElEmbedding
2. MuMu_data_2017_CMSSW923p2_MuEmbedding
3. MuTau_data_2017_CMSSW923p2_TauEmbedding
For testing purposes, you can copy a RAW testfile (/DoubleMuon/Run2017B-v1/RAW) with the name **test.root** into the respective folder and run:
`cmsRun selection.py`
`cmsRun lheprodandcleaning.py`
`cmsRun generator.py`
`cmsRun merging.py`
Final output will be a root file in MINIAOD format with the hybrid events.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by perrozzi** [comment]
+1

File: TauAnalysis/MCEmbeddingTools/plugins/CollectionMerger.cc [hunk 1]
```diff
@@ -71,21 +71,21 @@ void  CollectionMerger<T1,T2>::fill_output_obj_tracker(std::unique_ptr<MergeColl
...
-    DetId detIdObject( recHit->detid().rawId() );
-    T2 *akt_calo_obj = &output_map[detIdObject.rawId()];
-    float new_energy = akt_calo_obj->energy() + recHit->energy();
-    T2 newRecHit(*recHit);
-    newRecHit.setEnergy(new_energy);
-    *akt_calo_obj = newRecHit;
-  } 
+  std::map<uint32_t, BaseHit >   output_map;
+  // First merge the two collections again
+  for (auto inputCollection : inputCollections){
```
Comment (davidlange6) [inline TauAnalysis/MCEmbeddingTools/plugins/CollectionMerger.cc]: hi @perahrens - this can be auto const & inputCollection I believe?

Comment (perahrens) [inline TauAnalysis/MCEmbeddingTools/plugins/CollectionMerger.cc]: @davidlange6 yes thanks, the inputCollections are only read. I will add this commit to the PR.

**Comment by kpedro88** [comment]
-1
don't use cms-merge-topic to update a PR
http://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by perahrens** [comment]
I just wanted to add this commit b479f067bcaa092fb3604c7ba16b27bbca316cd0 to the PR. But now I did something wrong. Is there a command to revert this?

**Comment by kpedro88** [comment]
* make a backup branch of this one (that includes your new commit)
* restore the unmerged state of the branch
* follow the tutorial I linked above if necessary:
  * you only need to rebase if a) there's a merge conflict or b) you want to move your branch to a more recent IB (if the base of the branch was an IB that has expired)
  * otherwise, if you just need to recreate your working area (the base of your branch is a release that still exists), you just need to use `cms-checkout-topic`

Example commands for the first few bullet points:
```
cd [working area where you did cms-merge-topic]
git branch -m CMSSW_9_3_X CMSSW_9_3_X_backup
git push my-cmssw CMSSW_9_3_X_backup
git branch -m perahrens/CMSSW_9_3_X CMSSW_9_3_X
git push -f my-cmssw CMSSW_9_3_X
```

**Comment by kpedro88** [comment]
(and then you can cherry-pick your new commit from the backup branch)

**Comment by perahrens** [comment]
@kpedro88 Thank you very much for your support, Now it should be ok I guess.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by perrozzi** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Pull request merged**